sizeadjustingtextview/src/main/java/com/omitneedlesscode/sizeadjustingtextview/SizeAdjustingTextView.java

```diff
@@ -0,0 +1,172 @@
+        init();
+    }
+
+    @Override
+    public void onTextUpdated(String s, int start, int before, int after) {
+        // Since this view may be reused, it is good to reset the text size
+        resetTextSize();
+    }
+
+    private void init() {
+        mTextPaint = new Paint();
+        mTextPaint.setMultipleLine(true);
+        mTextPaint.setTextSize((int) mTextSize);
+        mTextPaint.setTextAlign(TextAlignment.CENTER);
+    }
+
+    @Override
+    public void setText(String text) {
+        super.setText(text);
+        resetTextSize();
+        resizeText();
+    }
+
+    @Override
+    public void setTextSize(int size) {
+        super.setTextSize(size);
+    }
+
+    @Override
+    public void setTextSize(int size, TextSizeType textSizeType) {
+        super.setTextSize(size, textSizeType);
+    }
+
+    public void setMaxTextSize(float maxTextSize) {
+        mMaxTextSize = maxTextSize;
+        postLayout();
+        invalidate();
+    }
+
+    public float getMaxTextSize() {
+        return mMaxTextSize;
+    }
+
+    public void setMinTextSize(float minTextSize) {
+        mMinTextSize = minTextSize;
+        postLayout();
+        invalidate();
+    }
+
+    public float getMinTextSize() {
+        return mMinTextSize;
+    }
+
+    public void resetTextSize() {
+        if (mTextSize > 0) {
+            super.setTextSize((int) mTextSize, TextSizeType.PX);
+            mMaxTextSize = mTextSize;
+        }
+    }
+
+    private void resizeText() {
+        int heightLimit = getHeight() - getPaddingBottom() - getPaddingTop();
+        int widthLimit = getWidth() - getPaddingLeft() - getPaddingRight();
+        resizeText(widthLimit, heightLimit);
+    }
+
+    public void resizeText(int width, int height) {
+        CharSequence text = getText();
+        if (text == null || text.length() == 0 || height <= 0 || width <= 0 || mTextSize == 0) {
+            return;
+        }
+        float newTextSize = findNewTextSize(width, height, text);
+        changeTextSize(newTextSize);
+    }
+
+    private void changeTextSize(float newTextSize) {
+        setTextSize((int) newTextSize, TextSizeType.PX);
+    }
+
+    private float findNewTextSize(int width, int height, CharSequence text) {
+        float targetTextSize = getTextSize();
+        int textHeight = getTextHeight(text, width, height, targetTextSize);
+        while (textHeight > height && targetTextSize > mMinTextSize) {
+            targetTextSize = Math.max(targetTextSize - 1, mMinTextSize);
+            textHeight = getTextHeight(text, width, height, targetTextSize);
+        }
+        return targetTextSize;
+    }
+
+    private int getTextHeight(CharSequence source, int width, int height, float textSize) {
+        mTextPaint.setTextSize((int) textSize);
+        mTextLayout = new SimpleTextLayout(source.toString(), mTextPaint,
+                new Rect(0, 0, width, height), width);
+        return mTextLayout.getHeight();
+    }
+}
```

Comment:
void setLineSpacing(float add, float mult)
boolean getAddEllipsis()
void setLineSpacing(float add, float mult)

These APIs are missing in the HMOS code. Please check.